[Release-1.5] Update security concept page for Istio 1.5 (#6727)
istio/istio.io: content/en/docs/concepts/security/index.md

    - /docs/concepts/policies/
---

Breaking down a monolithic application into atomic services offers various
benefits, including better agility, better scalability and better ability to
reuse services. However, microservices also have particular security needs:

- To defend against man-in-the-middle attacks, they need traffic encryption.

> - To provide flexible service access control, they need mutual TLS and

Reviewer: Should mutual TLS be in the glossary, or maybe a link here (first use)?

Author: @frankbu We should add mutual TLS to the glossary as a separate PR and add the shortcode to all pages that need it.

your services, wherever you run them. In particular, Istio security mitigates
both insider and external threats against your data, endpoints, communication,
and platform.

{{< image width="75%"
    link="./overview.svg"
    caption="Security overview"
    >}}

The Istio security features provide strong identity, powerful policy,
transparent TLS encryption, and authentication, authorization and audit (AAA)
tools to protect your services and data. The goals of Istio security are:

- Security by default: no changes needed to application code and
    infrastructure
- Defense in depth: integrate with existing security systems to provide
    multiple layers of defense
- Zero-trust network: build security solutions on distrusted networks

Visit our
[mutual TLS Migration docs](/docs/tasks/security/authentication/mtls-migration/)
to start using Istio security features with your deployed services. Visit our
[Security Tasks](/docs/tasks/security/) for detailed
instructions to use the security features.

## High-level architecture

Security in Istio involves multiple components:

- A Certificate Authority (CA) for key and certificate management

> - Istiod distributes to the proxies:

Reviewer: Istiod needs to be introduced before you can describe it's functions wrt security. That said, isn't Istiod too much an implementation detail for the concepts section? I thought one of the goals of the security concepts rewrite was to remove the architecture stuff (move it to the op guide) and focus instead on the security features. Like we did with traffic management.

Reviewer 2: Due to time constraints, we didn't think we'd have time to do a full rewrite.  Instead we focused on updating the documentation to correctly reflect the 1.5 architecture--and remove a few pieces of extraneous information.  We hope to have a rewrite of the Security Concepts page for 1.6.

Reviewer: Fair enough, but I still think we can try to keep this a little higher level. For example, my suggestion to use the generic Istio function (e.g., the config API server) instead of the implementation component name (Istiod) wherever possible.

Author: @frankbu In general I agree with you. The issue with using the generic function is that the audience won't necessarily know anything about it. Using Istiod in this case is better, because readers will recognize it from their deployment.

> - Istiod distributes to the proxies:

Reviewer: ```suggestion
- The configuration API server distributes to the proxies:
```

    - [authentication policies](/docs/concepts/security/#authentication-policies)
    - [authorization policies](/docs/concepts/security/#authorization-policies)
    - [secure naming information](/docs/concepts/security/#secure-naming)

> - Sidecar and perimeter proxies work as Policy Enforcement Points (PEPs) to

Reviewer: ```suggestion
- Sidecar and perimeter proxies work as [Policy Enforcement Points](link to what is it) (PEPs) to
```

Author: @frankbu I could use the following link: https://www.jerichosystems.com/technology/glossaryterms/policy_enforcement_point.html
WDYT?

   secure communication between clients and servers.
- A set of Envoy proxy extensions to manage telemetry and auditing

The control plane handles configuration from the API server and
configures the PEPs in the data plane. The PEPs are implemented using Envoy. The
following diagram shows the architecture.

{{< image width="75%"
    link="./arch-sec.svg"
    caption="Security Architecture"
    >}}

In the following sections, we introduce the Istio security features in detail.

## Istio identity

> Identity is a fundamental concept of any security infrastructure. At the

Reviewer: This section is inconsistent with the similar material in the glossary. In the glossary it talks about Istio assigning identifies to workloads but here it says they are given to service identities.
I suggest fixing this material to be consistent with the glossary version, and then shortening the glossary version to just have a short summary and a link to this section in the concepts.
https://preliminary.istio.io/docs/reference/glossary/#identity

Author: @frankbu I am changing the content in the section to use `workload` instead of `service`. That said, fully aligning all content to use the workload terminology is outside the scope of this PR. That also applies to reworking the identity glossary entry.

side, the server can determine what information the client can access based on
the
[authorization policies](/docs/concepts/security/#authorization-policies),
audit who accessed what at what time, charge clients based on the services they
used, and reject any clients who failed to pay their bill from accessing the
services.

In the Istio identity model, Istio uses the first-class service identity to
determine the identity of a service. This gives great flexibility and
granularity to represent a human user, an individual service, or a group of
services. On platforms that do not have such identity available, Istio can use
other identities that can group service instances, such as service names.

Use Istio service identities on different platforms:

- Kubernetes: Kubernetes service account
- GKE/GCE: GCP service account
- GCP: GCP service account
- AWS: AWS IAM user/role account
- On-premises (non-Kubernetes): user account, custom service account,
   service name, Istio service account, or GCP service account. The custom
   service account refers to the existing service account just like the
   identities that the customer's Identity Directory manages.

> ## PKI

Reviewer: ```suggestion
## PKI (Public Key Infrastructure)
```

The Istio PKI securely provisions strong identities to every workload with X.509
certificates. To automate key and certificate rotation at scale, the PKI runs an
Istio agent alongside each Envoy proxy for certificate and key provisioning. The
following diagram shows the identity provisioning flow.

{{< image width="75%"
    link="./id-prov.svg"
    caption="Identity Provision"
    >}}

Istio provisions identities through the secret discovery service (SDS) using the
following flow:

> 1. The CA offers a gRPC service to take CSR requests.

Reviewer: ```suggestion
1. The CA offers a gRPC service to take [certificate signing requests](https://en.wikipedia.org/wiki/Certificate_signing_request) (CSRs).
```

> 1. Envoy sends a certificate and key request via Envoy secret discovery

Reviewer: ```suggestion
1. Envoy sends a certificate and key request via the Envoy secret discovery
```

   service (SDS) API.
1. Upon receiving the SDS request, the Istio agent creates the private key
   and CSR before sending the CSR with its credentials to the Istio CA for signing.

> 1. The CAt validates the credentials carried in the CSR and signs the CSR to

Reviewer: ```suggestion
1. The CA validates the credentials carried in the CSR and signs the CSR to
```

the
[authorization policies](/docs/concepts/security/#authorization-policies),
audit who accessed what at what time, charge clients based on the workloads they
used, and reject any clients who failed to pay their bill from accessing the
workloads.

The Istio identity model uses the first-class `service identity` to
determine the identity of a request's origin. This model allows for great
flexibility and granularity for service identities to represent a human user, an
individual workload, or a group of workloads. On platforms without a service
identity, Istio can use other identities that can group workload
instances, such as service names.

The following list shows examples of service identities that you can use on different
platforms:

- Kubernetes: Kubernetes service account
- GKE/GCE: GCP service account
- GCP: GCP service account
- AWS: AWS IAM user/role account
- On-premises (non-Kubernetes): user account, custom service account,
   service name, Istio service account, or GCP service account. The custom
   service account refers to the existing service account just like the
   identities that the customer's Identity Directory manages.

> ## Public Key Infrastructure

Reviewer: ```suggestion
## Public Key Infrastructure (PKI) {#pki}
```

Reviewer: I actually wonder if the parens should be reversed, since PKI is the commonly used term??
```suggestion
## PKI (Public Key Infrastructure) {#pki}
```

> The Istio Public Key Infrastructure (PKI) securely provisions strong identities

Reviewer: ```suggestion
The Istio PKI securely provisions strong identities
```

traffic. It can also protect TCP traffic from general network hijackings.
However, secure naming doesn't protect from DNS spoofing because in that case
attackers hijack the DNS and modify the IP address of the destination. This is
because TCP traffic does not contain the hostname information and we can only
rely on the IP address for routing. In fact, this DNS hijack can happen even
before the client-side Envoy receives the traffic.

### Authentication architecture

You can specify authentication requirements for workloads receiving requests in
an Istio mesh using peer and request authentication policies. The mesh operator
uses `.yaml` files to specify the policies. The policies are saved in the Istio
configuration storage once deployed. The Istio controller watches the
configuration storage.

Upon any policy changes, the new policy is translated to the appropriate
configuration telling the PEP how to perform the required authentication
mechanisms. The control plane may fetch the public key and attach it to the
configuration for JWT validation. Alternatively, Istiod provides the path to the
keys and certificates the Istio system manages and installs them to the

> application pod for mutual TLS. You can find more info in the [PKI section](/docs/concepts/security/#public-key-infrastructure).

Reviewer: ```suggestion
application pod for mutual TLS. You can find more info in the [PKI section](/docs/concepts/security/#pki).
```

#### Value matching

Most fields in authorization policies support all the following matching
schemas:

- Exact match: exact string match.

> - Prefix match: A string with an ending `"*"`. For example, `"test.abc.*"`

Reviewer: ```suggestion
- Prefix match: a string with an ending `"*"`. For example, `"test.abc.*"`
```

   matches `"test.abc.com"`, `"test.abc.com.cn"`, `"test.abc.org"`, etc.

> - Suffix match: A string with a starting `"*"`. For example, `"*.abc.com"`

Reviewer: ```suggestion
- Suffix match: a string with a starting `"*"`. For example, `"*.abc.com"`
```

   matches `"eng.abc.com"`, `"test.eng.abc.com"`, etc.
- Presence match: `*` is used to specify anything but not empty. To specify
   that a field must be present, use the `fieldname: ["*"]`format. This is
   different from leaving a field unspecified, which means match anything,
   including empty.

There are a few exceptions. For example, the following fields only support exact
match:

- The `key` field under the `when` section
- The `ipBlocks` under the `source` section
- The `ports` field under the `to` section

> The following example policy allows access at paths with the  `/test/*` prefix

Reviewer: ```suggestion
The following example policy allows access at paths with the `/test/*` prefix
```

  name: allow-all
  namespace: default
spec:
  action: ALLOW
  rules:
  - {}
{{< /text >}}

> The example following example shows a simple  `deny-all` policy that denies any

Reviewer 3: Given we have "DENY" semantic now, should we rewrite this sentence to avoid confusion. E.g "The example following example shows a policy that does not allow any access to all workloads in the `admin` namespace."

May also show the recommendation and example to use DENY policy for such use case

+ @yangminzhu

Author: @diemtvu Thanks for pointing it out. I also fixed the duplicated example.

Reviewer 4: Yes, I think that makes sense:

We can change the current one to:
```
The following example shows a policy that does not allow any access to all workloads in the admin namespace. You can add more allow policies to allow extra access later.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: allow-nothing
  namespace: admin
spec:
  action: ALLOW
  rules:
  - {}
{{< /text >}}

The following example shows a policy that explicitly denies any access to all workloads in the `database` namespace. You can't add allow policies to allow extra access because the deny policy takes precedence over allow policy, meaning it will deny the request before any allow policies.

{{< text yaml >}}
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: deny-all
  namespace: database
spec:
  action: DENY
  rules:
  - {}
{{< /text >}}

Reviewer 4: @rcaballeromx I just made the comment, could you update with the latest change? thanks.

Author: @yangminzhu The first example yaml  has an `ALLOW` action. How can a reader know that it is denying access?

Reviewer 4: I just fixed a minor issue in my comment. The `allow-nothing` doesn't have `rules` which means it will never match, which means it allows nothing and any request will be denied.

Author: @yangminzhu Thanks for the correction. It is still confusing that readers need to use an allow policy to deny any requests. It is also not clear what the difference between `allow-nothing` and `deny-all` is. I can see how they both could be used depending on preference, but do we really want to introduce folks to both on the concept page? I would rather only keep the `deny-all` policy and keep things simple. @justinpettit WDYT?

Reviewer 2: I agree that we're adding complexity to a document that should probably be higher-level.  @yangminzhu, can we make a note of this and find a more appropriate location post-1.5?

Reviewer 4: @rcaballeromx the difference between `allow-nothing` and `deny-all` is the interaction between other policies. 

- If you have `allow-nothing`, it first restricts the permission to zero and you can incrementally add more allow policies to grant more permission. 

- If you have `deny-all`, all requests will be denied even if you add other allow policies, you need to delete the `deny-all` policy to allow other access.

We can keep the `deny-all` but please replace the current example with the one from my comment. That's a better way to express `deny-all` semantics.

Reviewer 4: @rcaballeromx This comment is not addressed in the merged PR, Rigs, would you mind to send out a follow PR to fix this? Thanks.
